misk-feature-testing/src/test/kotlin/misk/feature/testing/FakeFeatureFlagsTest.kt

```diff
@@ -89,6 +112,101 @@ internal class FakeFeatureFlagsTest {
+        .isEqualTo(JsonFeature("joker"))
+    assertThat(subject.getJson<JsonFeature>(FEATURE, "joker", badJokerAttributes))
+        .isEqualTo(JsonFeature("bad-joker"))
+    assertThat(subject.getJson<JsonFeature>(FEATURE, "joker", sleepyBadJokerAttributes))
+        .isEqualTo(JsonFeature("bad-joker"))
+    //Provides the key level override when there is no match on attributes
+    assertThat(subject.getJson<JsonFeature>(FEATURE, "joker", goodJokerAttributes))
+        .isEqualTo(JsonFeature("joker"))
+
+    subject.reset()
+    subject.override(FEATURE, JsonFeature("test-class"), JsonFeature::class.java)
+    subject.overrideKey(FEATURE, "joker", JsonFeature("test-key-class"), JsonFeature::class.java)
+    assertThat(subject.getJson<JsonFeature>(FEATURE))
+        .isEqualTo(JsonFeature("test-class"))
+    assertThat(subject.getJson<JsonFeature>(FEATURE, "joker"))
+        .isEqualTo(JsonFeature("test-key-class"))
+  }
+
+  @Test
+  fun getBoolean() {
+    // Default returns false and not throw as the other variants.
+    assertThat(subject.getBoolean(FEATURE, TOKEN)).isEqualTo(false)
+
+    // Can be overridden
+    subject.override(FEATURE, true)
+    subject.override(OTHER_FEATURE, false)
+    assertThat(subject.getBoolean(FEATURE, TOKEN)).isEqualTo(true)
+    assertThat(subject.getBoolean(OTHER_FEATURE, TOKEN)).isEqualTo(false)
+
+    // Can override with specific keys
+    subject.overrideKey(FEATURE, "joker", false)
+    assertThat(subject.getBoolean(FEATURE, TOKEN)).isEqualTo(true)
+    assertThat(subject.getBoolean(FEATURE, "joker")).isEqualTo(false)
+
+    // Can override with specific keys and attributes
+    val attributes = Attributes(mapOf("type" to "bad"))
+    subject.overrideKey(FEATURE, "joker", false, attributes)
+    assertThat(subject.getBoolean(FEATURE, TOKEN)).isEqualTo(true)
+    assertThat(subject.getBoolean(FEATURE, "joker")).isEqualTo(false)
+    assertThat(subject.getBoolean(FEATURE, "joker", attributes)).isEqualTo(false)
+    //Provides the key level override when there is no match on attributes
+    assertThat(subject.getBoolean(FEATURE, "joker", Attributes(mapOf("don't" to "exist"))))
+        .isEqualTo(false)
+  }
+
+  @Test
+  fun getString() {
+    // Default returns false and not throw as the other variants.
+    assertThrows<RuntimeException> { subject.getString(FEATURE, TOKEN) }
+
+    // Can be overridden
+    subject.override(FEATURE, "feature")
+    subject.override(OTHER_FEATURE, "other-feature")
+    assertThat(subject.getString(FEATURE, TOKEN)).isEqualTo("feature")
+    assertThat(subject.getString(OTHER_FEATURE, TOKEN)).isEqualTo("other-feature")
+
+    // Can override with specific keys
+    subject.overrideKey(FEATURE, "joker", "feature-joker")
+    assertThat(subject.getString(FEATURE, TOKEN)).isEqualTo("feature")
+    assertThat(subject.getString(FEATURE, "joker")).isEqualTo("feature-joker")
+
+    // Can override with specific keys and attributes
```

Comment:
nit: is it useful to test the override precedence for every single type? This test might have been simpler to have a bunch of tests for override precedence, and a bunch of tests for types, but the cross-product is not very useful.

Reply:
(this lousy structure was here when you got here, so you can continue it, but I think it’s bad)

Reply:
> nit: is it useful to test the override precedence for every single type?

it was useful from the test coverage perspective but also to uncover certain bugs when I did the refactoring.

> This test might have been simpler to have a bunch of tests for override precedence, and a bunch of tests for types, but the cross-product is not very useful.

Are you saying that we should organize the tests by having `override`, `overrideKey`, and `overrideKeyWithArguments`? each one of these methods will implement the logic for every single type?, something like:

```
fun override() {

  //override for Int
  //override for String
  //override for Enum
  //override for Boolean
  //override for Json
}
```

Happy to improve the test organization.